// Copyright 2024 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package wfa.measurement.reporting.v2alpha;

import "google/api/annotations.proto";
import "google/api/client.proto";
import "google/api/field_behavior.proto";
import "google/api/resource.proto";
import "google/protobuf/timestamp.proto";
import "wfa/measurement/reporting/v2alpha/basic_report.proto";

option java_package = "org.wfanet.measurement.reporting.v2alpha";
option java_multiple_files = true;
option java_outer_classname = "BasicReportsServiceProto";

// Service for managing 'BasicReport' resources.
service BasicReports {
  // Returns the 'BasicReport' with the given resource key.
  rpc GetBasicReport(GetBasicReportRequest) returns (BasicReport) {
    option (google.api.http) = {
      get: "/v2alpha/{name=measurementConsumers/*/basicReports/*}"
    };
    option (google.api.method_signature) = "name";
  }

  // Ordered by `create_time` ascending, `name` ascending.
  rpc ListBasicReports(ListBasicReportsRequest)
      returns (ListBasicReportsResponse) {
    option (google.api.http) = {
      get: "/v2alpha/{parent=measurementConsumers/*}/basicReports"
    };
    option (google.api.method_signature) = "parent";
  }
}

// Request mesage for 'GetBasicReport' method
message GetBasicReportRequest {
  // The name of the basic report to retrieve.
  string name = 1 [(google.api.field_behavior) = REQUIRED];
}

// Request mesage for 'GetListBasicReport' method
message ListBasicReportsRequest {
  // Message for specifying filters for listing BasicReports.
  message Filter {
    // Return BasicReports that were created after this time.
    google.protobuf.Timestamp created_after = 1
        [(google.api.field_behavior) = REQUIRED];
  }
  // Filter for the list of BasicReports
  Filter filter = 1 [(google.api.field_behavior) = OPTIONAL];

  // The maximum number of basic reports to return. The service may return fewer
  // than this value.
  // If unspecified, at most 10 basic reports will be returned.
  // The maximum value is 25; values above 25 will be coerced to 25.
  int32 page_size = 2 [(google.api.field_behavior) = OPTIONAL];

  // A page token, received from a previous `ListBasicReports` call.
  // Provide this to retrieve the subsequent page.
  //
  // When paginating, all other parameters provided to `ListBasicReports` must
  // match the call that provided the page token.
  string page_token = 3 [(google.api.field_behavior) = OPTIONAL];
}

// Response message for 'ListBasicReports' method.
message ListBasicReportsResponse {
  repeated BasicReport basic_reports = 1
      [(google.api.field_behavior) = OPTIONAL];

  // A token, which can be sent as `page_token` to retrieve the next page.
  // If this field is omitted, there are no subsequent pages.
  string next_page_token = 2 [(google.api.field_behavior) = OPTIONAL];
}
